### DISCUSSION TITLE
Combine safe selector + client extensions

### DISCUSSION THREAD

**Discussion by yovanoc** [prisma/prisma#18903] [Q&A] [answered]
Hello I have a question, maybe someone of you has a good pattern to do this but I use client extensions to add custom fields inside models. But I use this kind of helper to have safe selectors but with this pattern we didn't see our custom fields, how can we achieve something like that? thanks !

```ts
import type { Prisma, CheckSelectKeys } from '@prisma/client';

const createMyModelSelect = <
  T extends Prisma.MyModelSelect
>(
  arg: CheckSelectKeys<T, Prisma.MyModelSelect>
): CheckSelectKeys<T, Prisma.MyModelSelect> => arg;

export const selectMyModel = createMyModelSelect({
  id: true,
  // some fields
});

export type MyModel = Prisma.MyModelGetPayload<{
  select: typeof selectMyModel;
}>;
```

**Comment by ludralph** [comment]
Hi @yovanoc 👋 

I've shared your question internally with the team. One of our engineers will reach out to respond to this.

**Reply by Suniron** [reply]
Hey, did you have any news about your engineers? 🥲

I'm stuck with this problem too.

**Comment by yovanoc** [comment]
Oh big thanks!

Maybe there is a way to export types from a new "client" with extensions but didn't find it.

**Comment by millsp** [comment, marked as answer]
Hey, yes, this is possible with Client Extensions. We ship `Prisma.Args`, `Prisma.Result` and `Prisma.Payload` to assist you with that. Here's a very generic example that I took from our tests.

```ts
const xprisma = prisma.$extends({
    model: {
        $allModels: {
            findFirstOrCreate<T, const A extends Prisma.Args<T, 'findUniqueOrThrow'>>(
                this: T,
                _args: A,
            ): Prisma.Result<T, A, 'findUniqueOrThrow'> & { extra: boolean } {
                return {} as any
            },
        },
    },
})
```

In your case, and with what you're trying to achieve, it is also possible. First, you'll need to make sure that you can export your extended client. It will look slightly different to your first example, we will only rely on the new generic type tools discussed above:

```ts
const xprisma = prisma
// your previous extensions
.$extends({
    model: {
        $allModels: {
            createSelect<T, const A extends Prisma.Args<T, 'findUniqueOrThrow'>['select']>(this: T, args: A): A {
                return args
            },
        },
    },
})

const selectUser = xprisma.user.createSelect({
    name2: true,
})

export type UserSelection = Prisma.Result<typeof xprisma.user, { select: typeof selectUser }, 'findUniqueOrThrow'>
```

Let me know if that helps.

**Reply by yovanoc** [reply]
Oh ! This is a really good idea ! love this, gonna try it out soon!

**Comment by yovanoc** [comment]
Ok this is very good, I have a suggestion: findMany is not very accurate for my exemple because I need to add `[number]` to the end of every type to get only one.

**Reply by millsp** [reply]
Hey @yovanoc, I updated the example to use `findUniqueOrThrow`, which gives the purest result.
